contracts/TspLPPooling.sol

```diff
@@ -0,0 +1,259 @@
+
+        TspLP.transferFrom(msg.sender, address(this), _amount);
+
+        uint256 _shareAcc = _calculateCurrentAcc();
+
+        if(!delegators[msg.sender].hasDeposited){
+            delegators[msg.sender].hasDeposited = true;
+            delegators[msg.sender].availablePeanuts = 0;
+            delegators[msg.sender].debtRewards = _amount.mul(_shareAcc).div(1e12);
+            delegators[msg.sender].tspLPAmount = _amount;
+            delegatorsList.push(msg.sender);
+            return;
+        }
+        if (delegators[msg.sender].tspLPAmount > 0){
+            uint256 pending = delegators[msg.sender].tspLPAmount.mul(_shareAcc).div(1e12).sub(delegators[msg.sender].debtRewards);
+            if (pending > 0){
+                delegators[msg.sender].availablePeanuts = delegators[msg.sender].availablePeanuts.add(pending);
+            }
+        }
+        delegators[msg.sender].tspLPAmount = delegators[msg.sender].tspLPAmount.add(_amount);
+        totalDepositedTSPLP = totalDepositedTSPLP.add(_amount);
+
+        delegators[msg.sender].debtRewards = delegators[msg.sender].tspLPAmount.mul(_shareAcc).div(1e12);
+
+        emit DepositTSP(msg.sender, _amount);
+
+    }
+
+    function withdraw(uint256 _amount)
+        public
+        onlyDelegator
+    {
+        if (_amount == 0) return;
+
+        if (delegators[msg.sender].tspLPAmount == 0) return;
+
+        uint256 _shareAcc = _calculateCurrentAcc();
+        uint256 pending = delegators[msg.sender].tspLPAmount.mul(_shareAcc).div(1e12).sub(delegators[msg.sender].debtRewards);
+        if(pending > 0){
+            delegators[msg.sender].availablePeanuts = delegators[msg.sender].availablePeanuts.add(pending);
+        }
+        uint256 withdrawAmount;
+        if (_amount >= delegators[msg.sender].tspLPAmount)
+            withdrawAmount = delegators[msg.sender].tspLPAmount;
+        else
+            withdrawAmount = _amount;
+
+        require(withdrawAmount <= TspLP.balanceOf(address(this)), "ERC20: transfer amount exceeds balance");
+        TspLP.transfer(msg.sender, withdrawAmount);
+        delegators[msg.sender].tspLPAmount = delegators[msg.sender].tspLPAmount.sub(withdrawAmount);
+        totalDepositedTSPLP = totalDepositedTSPLP.sub(withdrawAmount);
+        delegators[msg.sender].debtRewards = delegators[msg.sender].tspLPAmount.mul(_shareAcc).div(1e12);
+
+
+        emit WithdrawTSP(msg.sender, _amount);
+
+    }
+
+    function withdrawPeanuts()
+        public
+        onlyDelegator
+    {
+        uint256 _shareAcc = _calculateCurrentAcc();
+
+        uint256 pending = delegators[msg.sender].tspLPAmount.mul(_shareAcc).div(1e12).sub(delegators[msg.sender].debtRewards);
+        if (pending > 0)
+            delegators[msg.sender].availablePeanuts = delegators[msg.sender].availablePeanuts.add(pending);
+        
+        require(delegators[msg.sender].availablePeanuts <= Pnuts.balanceOf(address(this)), "ERC20: transfer amount exceeds balance");
```

Comment:
If we can bump an EVENT and return rather than revert this transaction so that daemon can catch this situation and then trigger TSPPool contract update(for example deposit 1 sp, making reward distribute to Dev address), or if no user trigger TSPPool contract for a long time, we'd hang up here all the time.